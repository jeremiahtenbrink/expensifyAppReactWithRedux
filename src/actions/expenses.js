import uuid from "uuid";
import database from '../firebase/firebase';
//ADD EXPENSE ACTION
export const addExpense = (expense) => ({
	type: 'ADD EXPENSE',
	expense
});

export const startAddExpense = (expenseData = {}) => {
	return (dispatch, getState) => {
		const uid = getState().auth.uid;
		const {
			description = "",
			note = "",
			amount = 0,
			createdAt = 0,
		} = expenseData;
		const expense = { description, note, amount, createdAt };
		
		return database.ref(`users/${uid}/expenses`).push(expense).then((ref) => {
			dispatch(addExpense({
			    id: ref.key,
			    ...expense
		    }));
		});
	}
};

//EDIT EXPENSE ACTION
export const editExpense = (
	{
		id,
		updates,
	} = {} ) => ( {
	type: "EDIT EXPENSE",
	id,
	updates,
} );

export const startEditExpense = ({ id, updates} = {}) => {
	return( dispatch, getState) => {
		const uid = getState().auth.uid;
		console.log(uid);
		return database.ref( `users/${uid}/expenses/${ id }` ).update( updates ).then( () => {
			dispatch(editExpense({id,updates}));
		} ).catch((e) => {
		    console.log(e);
		});
		
	}
};

// REMOVE EXPENSE BY ID
export const removeExpense = ( { id = undefined } = {} ) => ( {
	type: "REMOVE EXPENSE",
	id,
} );

export const startRemoveExpense = (idObject) => {
	return (dispatch, getState) => {
		const uid = getState().auth.uid
		return database.ref(`users/${uid}/expenses/${idObject.id}`).remove().then(() => {
			
			dispatch( removeExpense(idObject));
			
		}).catch(function(error) {
			console.log("Remove failed: " + error.message)
		});
	};
};

//SET EXPENSES
export const setExpenses = (expenses) => ({
    type: 'SET EXPENSES',
	expenses
});

export const startSetExpenses = () => {
    return (dispatch, getState) => {
	    const uid = getState().auth.uid;
    	return database.ref(`users/${uid}/expenses`).once('value').then((snapshot) => {
    	    const expenses = [];
    	    snapshot.forEach((childSnapshot) => {
    	        expenses.push({
		            id: childSnapshot.key,
		            ...childSnapshot.val()
	            });
    	    });
    	    dispatch(setExpenses(expenses));
    	});
    };
};



