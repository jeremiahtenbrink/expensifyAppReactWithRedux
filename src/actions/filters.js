//SET TEXT FILTER
export const setTextFilter = (text = '') => ({
	type: "SET FILTER TEXT",
	text,
});

//SORT BY AMOUNT
export const sortByAmount = () => ({
    type: 'SORT BY AMOUNT'
});

//SORT BY DATE
export const sortByDate = () => ({
    type: 'SORT BY DATE'
});


//SET START DATE
export const setStartDate = (date = undefined) => ({
    type: "SET START DATE",
    date
});


//SET END DATE
export const setEndDate = (date = undefined) => ({
    type: "SET END DATE",
    date
});
