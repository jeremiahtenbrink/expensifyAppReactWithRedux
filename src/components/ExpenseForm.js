import React from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";
import { connect } from "react-redux";

export class ExpenseForm extends React.Component {
	constructor ( props )
	{
		super( props );
		
		this.state = {
			description    : props.expense ? props.expense.description : "",
			note           : props.expense ? props.expense.note : "",
			amount         : props.expense ? ( props.expense.amount / 100 ).toString() : "",
			createdAt      : props.expense ? moment( props.expense.createdAt ) : moment(),
			calendarFocused: false,
			errorMessage   : "",
		};
	}
	
	onDescriptionChange = ( e ) => {
		const description = e.target.value;
		this.setState( () => ( { description } ) );
	};
	
	onTextAreaChange = ( e ) => {
		const note = e.target.value;
		this.setState( () => ( { note } ) );
	};
	
	onAmountChange = ( e ) => {
		const amount = e.target.value;
		if ( !amount || amount.match( /^\d{1,}(\.\d{0,2})?$/ ) ){
			this.setState( () => ( { amount } ) );
		}
	};
	
	onDateChange = ( createdAt ) => {
		if ( createdAt ){
			this.setState( () => ( { createdAt } ) );
		}
	};
	
	onFocuseChange = ( { focused } ) => {
		this.setState( () => ( { calendarFocused: focused } ) );
	};
	
	onSubmit = ( e ) => {
		e.preventDefault();
		
		if ( !this.state.description || !this.state.amount ){
			// set error state
			this.setState( () => ( { errorMessage: "Please provide description and amount." } ) );
		} else{
			this.setState( () => ( { errorMessage: "" } ) );
			// send to edit or add expense page
			this.props.onSubmit( {
				description: this.state.description,
				amount     : parseFloat( this.state.amount ) * 100,
				createdAt  : this.state.createdAt.valueOf(),
				note       : this.state.note,
			} );
		}
	};
	
	render ()
	{
		return (
			<div>
				{ this.state.errorMessage && <p>{ this.state.errorMessage }</p> }
				<form onSubmit={ this.onSubmit }>
					<input type="text"
					       placeholder="Description"
					       autoFocus
					       value={ this.state.description }
					       onChange={ this.onDescriptionChange }
					/>
					<input type="text"
					       placeholder="Amount"
					       value={ this.state.amount }
					       onChange={ this.onAmountChange }
					/>
					<SingleDatePicker
						date={ this.state.createdAt }
						onDateChange={ this.onDateChange }
						focused={ this.state.calendarFocused }
						onFocusChange={ this.onFocuseChange }
						numberOfMonths={ 1 }
						isOutsideRange={ () => false }
					/>
					<textarea name="note" id="" cols="20" rows="2"
					          value={ this.state.note }
					          placeholder="Add a note for your expense"
					          onChange={ this.onTextAreaChange }/>
					<button>Add Expense</button>
				
				</form>
			</div>
		);
	}
}

export default connect()( ExpenseForm );
