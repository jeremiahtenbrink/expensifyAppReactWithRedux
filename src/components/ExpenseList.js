import React from "react";
import PropTypes from "prop-types";
import { connect } from "react-redux";
import ExspenseListItem from "./ExspenseListItem";
import selectExpenses from "../selectors/expenses";

export const ExpenseList = ( props ) => (
	<div>
		{
			props.expenses.length === 0 ? (
				<p>No expenses</p>
			): (
				props.expenses.map( ( expense ) => {
					return <ExspenseListItem key={ expense.id } { ...expense } />;
				} )
			)
		}
		
	</div>
);

const mapStateToProps = state => ( {
	expenses: selectExpenses( state.expenses, state.filter ),
} );

export default connect( mapStateToProps )( ExpenseList );

ExpenseList.propTypes = {};
ExpenseList.defaultProps = {};


