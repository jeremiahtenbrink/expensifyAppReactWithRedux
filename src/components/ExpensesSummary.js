import React from "react";
import numeral from "numeral";
import selectExpenses from "../selectors/expenses";
import { getExpensesTotal } from "../selectors/expenses-total";
import { connect} from "react-redux";

export const ExpensesSummary = ( { expenseCount, expensesTotal, totalNumberOfExpenses } ) => {
	
	const expenseWord = expenseCount === 1 ? "expense" : "expenses";
	const formattedExpensesTotal = numeral( expensesTotal / 100 ).format( "$0,0.00" );
	
	return (
		<div>
			<h3>
				Showing { expenseCount } of { totalNumberOfExpenses } { expenseWord }:
				
			</h3>
			<h3>
				Totalling: { formattedExpensesTotal }
			</h3>
		</div>
	);
};

const mapStateToProps = ( state ) => {
	const visibleExpenses = selectExpenses( state.expenses, state.filter );
	
	return {
		expenseCount        : visibleExpenses.length === 0 ? 0 : visibleExpenses.length,
		expensesTotal        : getExpensesTotal( visibleExpenses ),
		totalNumberOfExpenses: state.expenses.length,
	};
};

export default connect( mapStateToProps )( ExpensesSummary );


