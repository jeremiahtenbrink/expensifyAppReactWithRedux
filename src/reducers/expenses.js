const expensesReducerDefaultState = [];

export default ( state = expensesReducerDefaultState, action ) => {
	switch ( action.type ){
		case "ADD EXPENSE":
			return [
				...state,
				action.expense,
			];
		case "REMOVE EXPENSE":
			return state.filter( ( expense ) => expense.id !== action.id );
		case "EDIT EXPENSE":
			return state.map( ( expense ) => {
				if ( expense.id === action.id ){
					return { ...expense, ...action.updates };
					
				} else{
					return expense;
				}
			} );
		case "SET EXPENSES":
			return action.expenses;
		default:
			return state;
	}
};

