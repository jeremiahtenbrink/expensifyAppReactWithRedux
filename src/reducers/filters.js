// FILTER REDUCER

import moment from "moment";

const filterReducerDefaultState = ( {
	text     : "",
	sortBy   : "date",
	startDate: moment().startOf( "month" ),
	endDate  : moment().endOf( "month" ),
} );

export default ( state = filterReducerDefaultState, action ) => {
	switch ( action.type ){
		case "SET FILTER TEXT":
			return { ...state, text: action.text };
		case "SORT BY AMOUNT":
			return { ...state, sortBy: "amount" };
		case "SORT BY DATE":
			return { ...state, sortBy: "date" };
		case "SET START DATE":
			return { ...state, startDate: action.date };
		case "SET END DATE":
			return { ...state, endDate: action.date };
		default:
			return state;
	}
};