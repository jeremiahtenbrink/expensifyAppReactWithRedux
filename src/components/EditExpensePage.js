import React, { Component } from "react";
import ExpenseForm from "./ExpenseForm";
import { connect } from "react-redux";
import { editExpense, removeExpense } from "../actions/expenses";

export class EditExpensePage extends Component {
	
	onEditExpense = (expense) => {
		this.props.editExpense( {
			id: this.props.expense.id,
			updates: expense} );
		this.props.history.push( "/" );
	};
	
	onRemoveExpense = () => {
		
		this.props.removeExpense({ id: this.props.expense.id });
		this.props.history.push( "/" );
	};
	
	render ()
	{
		
		return (
			<div>
				<ExpenseForm id="expense-form"
					expense={ this.props.expense }
					onSubmit={this.onEditExpense}
				/>
				
				<button onClick={this.onRemoveExpense}>Remove
				</button>
			</div>
		
		);
	}
}

const mapStateToProps = ( state, props ) => {
	return {
		expense: state.expenses.find( ( expense ) => expense.id === props.match.params.id ),
	};
};

const mapDispatchToProps = (dispatch) => ({
	editExpense: (expense) => dispatch(editExpense(expense)),
	removeExpense: (idObject) => dispatch(removeExpense( idObject ))
});

export default connect( mapStateToProps, mapDispatchToProps )( EditExpensePage );