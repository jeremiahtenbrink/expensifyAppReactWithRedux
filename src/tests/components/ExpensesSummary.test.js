import React from "react";
import { ExpensesSummary } from "../../components/ExpensesSummary";
import { shallow } from "enzyme/build";
import expenses from "../fixtures/expenses";
import {filters} from "../fixtures/filters";

test( "should render expense summary with 1 item", () => {
	const wrapper = shallow( <ExpensesSummary expenseCount={1} expensesTotal={400} totalNumberOfExpenses={1}  /> );
	expect( wrapper ).toMatchSnapshot();
} );

test( "should render expense summary with multiple expense items", () => {
	const wrapper = shallow( <ExpensesSummary expenseCount={3} expensesTotal={40060} totalNumberOfExpenses={5}/> );
	expect( wrapper ).toMatchSnapshot();
} );