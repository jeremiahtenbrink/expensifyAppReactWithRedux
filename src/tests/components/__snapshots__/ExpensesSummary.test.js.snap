// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense summary with 1 item 1`] = `
<div>
  <h3>
    Showing 
    1
     of 
    1
     
    expense
    :
  </h3>
  <h3>
    Totalling: 
    $4.00
  </h3>
</div>
`;

exports[`should render expense summary with multiple expense items 1`] = `
<div>
  <h3>
    Showing 
    3
     of 
    5
     
    expenses
    :
  </h3>
  <h3>
    Totalling: 
    $400.60
  </h3>
</div>
`;
